// ngx-linky/bundles/ngx-linky.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('autolinker')) :
    typeof define === 'function' && define.amd ? define('ngx-linky', ['exports', '@angular/core', 'autolinker'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global["ngx-linky"] = {}, global.ng.core, global.Autolinker));
})(this, (function (exports, i0, Autolinker) { 'use strict';

    function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

    function _interopNamespace(e) {
        if (e && e.__esModule) return e;
        var n = Object.create(null);
        if (e) {
            Object.keys(e).forEach(function (k) {
                if (k !== 'default') {
                    var d = Object.getOwnPropertyDescriptor(e, k);
                    Object.defineProperty(n, k, d.get ? d : {
                        enumerable: true,
                        get: function () { return e[k]; }
                    });
                }
            });
        }
        n["default"] = e;
        return Object.freeze(n);
    }

    var i0__namespace = /*#__PURE__*/_interopNamespace(i0);
    var Autolinker__default = /*#__PURE__*/_interopDefaultLegacy(Autolinker);

    var LinkyPipe = /** @class */ (function () {
        function LinkyPipe() {
        }
        LinkyPipe.prototype.transform = function (value, options) {
            return Autolinker__default["default"].link(value, options);
        };
        return LinkyPipe;
    }());
    LinkyPipe.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0__namespace, type: LinkyPipe, deps: [], target: i0__namespace.ɵɵFactoryTarget.Pipe });
    LinkyPipe.ɵpipe = i0__namespace.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0__namespace, type: LinkyPipe, name: "linky" });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0__namespace, type: LinkyPipe, decorators: [{
                type: i0.Pipe,
                args: [{ name: 'linky' }]
            }] });

    var LinkyModule = /** @class */ (function () {
        function LinkyModule() {
        }
        return LinkyModule;
    }());
    LinkyModule.ɵfac = i0__namespace.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0__namespace, type: LinkyModule, deps: [], target: i0__namespace.ɵɵFactoryTarget.NgModule });
    LinkyModule.ɵmod = i0__namespace.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0__namespace, type: LinkyModule, declarations: [LinkyPipe], exports: [LinkyPipe] });
    LinkyModule.ɵinj = i0__namespace.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0__namespace, type: LinkyModule });
    i0__namespace.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0__namespace, type: LinkyModule, decorators: [{
                type: i0.NgModule,
                args: [{
                        declarations: [LinkyPipe],
                        exports: [LinkyPipe],
                    }]
            }] });

    /*
     * Public API Surface of ngx-linky
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.LinkyModule = LinkyModule;
    exports.LinkyPipe = LinkyPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
