// ngx-linky/fesm2015/ngx-linky.js
import * as i0 from '@angular/core';
import { Pipe, NgModule } from '@angular/core';
import Autolinker from 'autolinker';

class LinkyPipe {
    transform(value, options) {
        return Autolinker.link(value, options);
    }
}
LinkyPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: LinkyPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
LinkyPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: LinkyPipe, name: "linky" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: LinkyPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'linky' }]
        }] });

class LinkyModule {
}
LinkyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: LinkyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LinkyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: LinkyModule, declarations: [LinkyPipe], exports: [LinkyPipe] });
LinkyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: LinkyModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.13", ngImport: i0, type: LinkyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LinkyPipe],
                    exports: [LinkyPipe],
                }]
        }] });

/*
 * Public API Surface of ngx-linky
 */

/**
 * Generated bundle index. Do not edit.
 */

export { LinkyModule, LinkyPipe };
